### PR TITLE
Feature/internationalize client components

### PR DESCRIPTION
**Pull request by miurahr**
- i18n with standard angular9 i18n module
- i18n-ed UI components
- Introduce XLIFF message catalog template
- Add Japanese and Germany sample translations
- Add localized targets

### CODE REVIEW

**Comment by miurahr** [comment]
You can test localized client by running `npm run dev-ja`.

**Comment by miurahr** [comment]
Now ready for review.

* This PR does not care about production deployment, and docker images.
* Tested only with local development build.

File: client/src/app/project/project-creation/project-creation.component.html [hunk 1]
```diff
@@ -1,34 +1,34 @@
...
-				<p>Description:</p>
+				<p i18n>Description:</p>
 				<textarea name="description-text-area" class="description-text-area" [(ngModel)]="projectDescription"></textarea>
 			</div>
 		</form>
 
-		<h2>Create shapes:</h2>
+		<h2 i18n>Create shapes:</h2>
 		<!-- Different shape creation methods: Divide, Upload -->
 		<app-tabs [tabs]="['Draw', 'Upload', 'Remote', 'Remove']" #tabs (tabSelected)="onTabSelected($event)">
```
Comment (miurahr) [inline client/src/app/project/project-creation/project-creation.component.html:33]: Here is a untranslated part of tabs.

File: client/src/app/project/project/project.component.html [hunk 2]
```diff
@@ -1,17 +1,17 @@
...
 
 <div class="root-container">
 	<div class="task-list-details-container">
 		<div class="description-container" *ngIf="project?.description.trim().length !== 0">
-			<h2 class="description-label">Description:</h2>
+			<h2 class="description-label" i18n="Description label">Description:</h2>
 			<div class="description-text">{{project?.description}}</div>
 		</div>
-		<app-tabs [tabs]="['Tasks', 'Users', 'Settings']" #tabs>
+		<app-tabs i18n-[tabs] [tabs]="['Tasks', 'Users', 'Settings']" #tabs>
```
Comment (miurahr) [inline client/src/app/project/project/project.component.html:14]: tabs translations are not shown on UI. help wanted.

**Comment by hauke96** [comment]
This is a great feature, which looks good so far. I pushed one commit making the angular dev-server work again.

The deployment and hosting is a larger issue as we get multiple separate applications for each language. As far as I read, nginx and apache are able to deal with browser locales but I don't know if and how well this works. So this is - as you mentioned - an open issue.

The tabs are interesting as well, I currently don't have any solution. But I'm sure we'll figure it out.

There are also texts within the typescript code (e.g. error messages), we need to translate them as well.

As soon as all the issues are fixed, I'd love to merge this. Unfortunately I don't have that much time right now, but we should keep working on this!

**Comment by miurahr** [comment]
> but I don't know if and how well this works.
I should work and my field, but I don't have enough time now.

> The tabs are interesting as well, I currently don't have any solution. But I'm sure we'll figure it out.
I'd like to keep it as pending.

> texts within the typescript code (e.g. error messages)
Added the changes.

Update npm build to produce configured three languages: en_US, ja, and de. a tree under dist/  is changed
to `dist/simple-task-manager/en_US/`  so please update deployment script.

File: client/src/app/translation-extraction.component.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
+/**
+ * Workaround for extraction of localize keys
```
Comment (miurahr) [inline client/src/app/translation-extraction.component.ts:2]: work around for extraction of localize keys

**Comment by miurahr** [comment]
Update  nginx config.

After building and launching stm-client-test docker image, when accessing http://localhost/ , it redirect to http://localhost/en-US/ 
When accessing http://localhost/de/   or /ja/ you see localized site.

File: client/src/app/translation-extraction.component.ts (hunk 3 above)
Comment (hauke96) [reply client/src/app/translation-extraction.component.ts:2]: Maybe we can do something like this?

`<span i18n="@@WARN_ALREADY_MEMBER">User {user} is already a member of this project</span>`

And then manually replacing the substring `{user}` by the username after translation? Haven't tested it yet and it'is even more code to write, but maybe that will work.

Comment (miurahr) [reply client/src/app/translation-extraction.component.ts:2]: Angular-CLI  `ng xi18n` genearete different id for `<span i18n="@@WARN_ALREADY_MEMBER">User {user} is already a member of this project</span>`  against `$localize`:@@MESSAGE_ID:User ${ user }:user: is already`

The farmer become such as
```
<source>User <x id="INTERPOLATION" alter="{user}"/> is already a member of this project</source>
```
in where the latter should be
```
<source>User <x id="PH"/> is already a member of this project</source>
```

It cause the record in message catalog become invalid, UI show a broken message such as `@@:MESSAGE_ID:User { George }:user: is already a member of this project` which include id and placeholder name, and logs error which said there is no interpolation placeholder in the text.

You can provide something `fixer` utility to convert placeholder tag which derived from typescript text into correct one, if you want to use it.

A blog describe behavior here: https://blog.ninja-squad.com/2019/12/10/angular-localize/

Comment (miurahr) [reply client/src/app/translation-extraction.component.ts:2]: A prospective place holder, INTERPOLATION,  is described here https://angular.io/guide/i18n#translate-plurals-and-alternate-expressions

Comment (miurahr) [reply client/src/app/translation-extraction.component.ts:2]: I found a way to hack this, using `$localize:@@MESSAGE_ID:User ${ user }:INTERPOLATION: is already` to force place holder id to "INTERPOLATION"

File: client/src/app/project/project-creation/project-creation.component.ts [hunk 4]
```diff
@@ -174,7 +174,7 @@ export class ProjectCreationComponent implements OnInit, AfterViewInit {
         this.router.navigate(['/manager']);
       }, e => {
         console.error(e);
-        this.notificationService.addError('Could not create project');
+        this.notificationService.addError($localize`:@@ERROR_NOT_CREATE_PROJ:Could not create project`);
```
Comment (miurahr) [inline client/src/app/project/project-creation/project-creation.component.ts:177]: xi18n hack  for here is lacked

File: client/nginx.conf [hunk 5]
```diff
@@ -15,10 +15,14 @@ server {
 	ssl_certificate /etc/letsencrypt/live/stm.hauke-stieler.de/cert.pem;
 	ssl_certificate_key /etc/letsencrypt/live/stm.hauke-stieler.de/privkey.pem;
 
-	location / {
+	location ~ ^/(en|ja|de)/ {
```
Comment (miurahr) [inline client/nginx.conf:18]: This should be `~ ^/en-US/ja/de/`

File: client/nginx.conf [hunk 6]
```diff
@@ -15,10 +15,14 @@ server {
...
 		root   /usr/share/nginx/html;
 		index  index.html;
 		try_files $uri $uri/ /index.html;
 	}
+
+	location / {
+		rewrite ^/(.*)$ /en/$1 last;
```
Comment (miurahr) [inline client/nginx.conf:25]: This should be `/en-US/`

File: client/nginx.conf (hunk 5 above)
Comment (miurahr) [reply client/nginx.conf:18]: When set here as `~ ^/en-US/` then only English is provided on the site. It depends on production site operation policy.

File: client/src/app/translation-extraction.component.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+import { ChangeDetectionStrategy, Component } from '@angular/core';
+
+@Component({
+    selector: 'translation-extraction',
+    template: `
+        <span i18n="@@WARN_AUTH_FAIL">Logout because authorization was not successful</span>
+        <span i18n="@@ERROR_NOT_CREATE_PROJ">Could not create project</span>
+        <span i18n="@@ERROR_LOAD_PROJECTS">Could not load projects '{{interp}}'</span>
+        <span i18n="@@ERROR_LIVE_UPDATE">Could not initialize live-updates</span>
+        <span i18n="@@ERROR_COULD_NOT_UPLOAD">Could not upload file '</span>
```
Comment (miurahr) [inline client/src/app/translation-extraction.component.ts:19]: fix missinf for interpolation

File: client/nginx.conf (hunk 5 above)
Comment (miurahr) [reply client/nginx.conf]: fix uploaded.

File: client/nginx.conf (hunk 6 above)
Comment (miurahr) [reply client/nginx.conf]: fix uploaded.

File: client/src/app/translation-extraction.component.ts (hunk 7 above)
Comment (miurahr) [reply client/src/app/translation-extraction.component.ts]: fixed.

**Comment by miurahr** [comment]
I think all the things except for translation of tabs are done.
Could @hauke96 review again?

File: client/src/app/user/user-invitation/user-invitation.component.ts [hunk 8]
```diff
@@ -25,7 +25,7 @@ export class UserInvitationComponent implements OnInit {
 
   public onInvitationButtonClicked() {
     if (this.users.map(u => u.name).includes(this.enteredUserName)) {
-      this.notificationService.addWarning('User ${ this.enteredUserName } ' + $localize`:@@WARN_ALREADY_MEMBER:is already a member of this project`);
+      this.notificationService.addWarning($localize`:@@WARN_ALREADY_MEMBER:User ${ this.enteredUserName }:INTERPOLATION:' is already a member of this project`);
```
Comment (hauke96) [inline client/src/app/user/user-invitation/user-invitation.component.ts:28]: I think you missed a ' here?
```
... User '${ this...
         ^
```

File: client/Dockerfile-test [hunk 9]
```diff
@@ -4,6 +4,7 @@ FROM simpletaskmanager/stm-client-base:1.1.0-dev as builder
 COPY ./ /stm-client/
 WORKDIR /stm-client/
 
+RUN NODE_OPTIONS="--max_old_space_size=4096" npm install @angular/localize
```
Comment (hauke96) [inline client/Dockerfile-test:7]: Why is this necessary? The @angular/localize package would be installed via the package.json, wouln't it?

**Comment by hauke96** [comment]
I made some fixes to the nginx and docker files and the deployment and hosting works (including the redirect to the en-US default locale). Please don't change the test configs as they are used for the test server https://stm-test.hauke-stieler.de.

I also translated the tabs: The tab-titles (e.g. "Tasks", "Users" and "Settings") is now provided by the typescript code where we can easily translate the string literals. This works quite nicely, even though some German words are longer than the English ones and therefore e.g. the project creation area is too small :D But that's a minor issue.

**Comment by miurahr** [comment]
A Germany translation is incomplete, so could you please check and review before merge?

**Comment by miurahr** [comment]
Tab localization select wrong language in test server.

スクリーンショット 2020-08-01 10-32-43

**Comment by miurahr** [comment]
- I've fixed two small problems; typo in message-extraction hack, and l10n for Open in iD button.
- Progress several messages in German, which is based on machine translation, where the MT engine permit me to translate OSS ([TexTra (license, appendix 1)](https://mt-auto-minhon-mlt.ucri.jgn-x.jp/content/policy/)by NICT, Japan government's lab. )

**Comment by hauke96** [comment]
Please review my changes. The german translation is complete and I fixed some styling issues due to longer words. I also updates the test server, so https://stm-test.hauke-stieler.de not hosts the latest version of this branch.

**Comment by miurahr** [comment]
I believe all things is done (except for language selector UI parts, for example [this tutorial](https://phrase.com/blog/posts/angular-9-tutorial-on-i18n/)).

**Comment by hauke96** [comment]
Thanks a lot for your contributions! I create an separate issue for the language selection but the i18n part itself is done.
